mat64/io.go

```diff
@@ -5,14 +5,152 @@
 package mat64
 
 import (
+	"bytes"
 	"encoding/binary"
+
+	"github.com/gonum/matrix"
 )
 
 var (
-	littleEndian  = binary.LittleEndian
-	bigEndian     = binary.BigEndian
-	defaultEndian = littleEndian
-
+	binLE       = binary.LittleEndian
 	sizeInt64   = binary.Size(int64(0))
 	sizeFloat64 = binary.Size(float64(0))
 )
+
+// MarshalBinary encodes the receiver into a binary form and returns the result.
+//
+// Dense is little-endian encoded as follows:
+//   0 -  8  number of rows    (int64)
+//   8 - 16  number of columns (int64)
+//  16 - ..  matrix data elements (float64)
+//           [0,0] [0,1] ... [0,ncols-1]
+//           [1,0] [1,1] ... [1,ncols-1]
+//           ...
+//           [nrows-1,0] ... [nrows-1,ncols-1]
+func (m Dense) MarshalBinary() ([]byte, error) {
+	buf := bytes.NewBuffer(make([]byte, 0, m.mat.Rows*m.mat.Cols*sizeFloat64+2*sizeInt64))
```

Comment:
If I have a large matrix on a 386 (16kx16k elements), I could conceivably cause a panic here because m.mat.Rows*m.mat.Cols*sizeFloat64+2*sizeInt64 > 2G. So check that m.mat.Rows*m.mat.Cols*sizeFloat64+2*sizeInt64 <= maxInt and return an error otherwise. Equivalent for Vector.

It's not going to catch much, but if we miss one that's bad.